### PR TITLE
Add `Metadata::ForceRemeshComm` Flag

### PR DESCRIPTION
**Pull request by lroberts36**
## PR Summary

Add a metadata flag `Metadata::RemeshComm` that guarantees that the values of a field will be communicated between ranks during remeshing. Fields with `Metadata::Independent` and `Metadata::FillGhost` are also automatically communicated, as was previously done, so this should not change the behavior of existing downstream codes. 

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Review by Yurlungur — APPROVED** [review]
LGTM

**Comment by Yurlungur** [comment]
@jdolence @pgrete could we get this reviewed and in? I need it for the prolongation/restriction stuff

**Comment by Yurlungur** [comment]
Ping @forrestglines @jdolence @pgrete this is a small change and needs approval

**Comment by lroberts36** [comment]
Ping @forrestglines @jdolence @pgrete for a review

File: docs/interface/Metadata.md
Comment (pgrete) [inline docs/interface/Metadata.md]: I'm not sure I follow the comment.
How can running on a different number of ranks produce different results?
I could image this happening when a variable is neither `independent` nor `dependent`, but think we should not allow for that case.

File: src/mesh/meshblock.cpp
Comment (pgrete) [inline src/mesh/meshblock.cpp:195]: Likely related to my question above: In which case you want to communicate a variable but not refine it?
I'm probably missing the big picture.

File: docs/interface/Metadata.md
Comment (lroberts36) [reply docs/interface/Metadata.md]: When a block is moved in the block list within a rank, all of the fields are copied over. When a block is moved between ranks, only independent data is transferred on the assumption that the dependent data can be reconstituted on the new rank from the dependent data. This is normally the case, but sometimes it may be the case that it is too expensive or introduces numerical noise to rebuild some `dependent` fields. Then, it is useful to have a flag that indicates the data in particular dependent variables should be communicated during load balancing.

File: src/mesh/meshblock.cpp
Comment (lroberts36) [reply src/mesh/meshblock.cpp:195]: We were running into issues in Riot where results differed when run on different numbers of ranks, which that seemed to be related to certain `dependent` variables not being initialized after blocks were moved between ranks. Presumably, by very carefully reconstituting the dependent variables, we could get bitwise agreement independent of number of ranks, but a simpler solution was to just communicate some dependent data. 

When refining, we still only want to refine `independent` variables.

File: docs/interface/Metadata.md
Comment (pgrete) [reply docs/interface/Metadata.md]: > When a block is moved in the block list within a rank, all of the fields are copied over. 

I thought we'd should move around the pointer to the block and update the data within the block and not copy the data itself.

> When a block is moved between ranks, only independent data is transferred on the assumption that the dependent data can be reconstituted on the new rank from the dependent data.

Yes, this is/was also my working assumption when I wrote the comment.

> This is normally the case, but sometimes it may be the case that it is too expensive or introduces numerical noise to rebuild some dependent fields. Then, it is useful to have a flag that indicates the data in particular dependent variables should be communicated during load balancing.

I'm probably a little hesitant as I'm afraid that the proposed functionality/fix hides a deeper problem and/or invites for some "unfavorable" design choices downstream.
While I understand that there's some extra cost involved in order to rebuild dependent data, I'm currently weighing the "strict" design decision that "everything derived is handled by `FillDerived` at clear points in the dataflow" as a stronger argument than performance (but probably your performance use case is vastly different from the ones I have in mind).

How about we discuss this tomorrow during the call? I'd like to hear feedback from more people on this as I consider this a breaking change (with respect to the underlying design decision).

Comment (lroberts36) [reply docs/interface/Metadata.md]: Sure, I am happy to discuss it during the call tomorrow.

**Comment by lroberts36** [comment]
@pgrete: I updated the documentation to reflect our discussion and renamed the flag to `ForceRemeshComm`. Can you please approve when you get the change?

**Review by pgrete — APPROVED** [review]

**Pull request merged**
